### PR TITLE
Azure Blob Ext_Pillar

### PR DESCRIPTION
**Pull request by Ajnbro**
Some of the descriptions of parameters will probably need to be tweaked for improved accuracy and/or greater detail.

### CODE REVIEW

File: salt/_pillar/azureblob.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
+# -*- coding: utf-8 -*-
+'''
+Use Azure Blob as a Pillar source
+
+.. versionadded:: Sodium
+
+:maintainer: <devops@eitr.tech>
+:maturity: new
+:depends:
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:9]: dupe depends

File: salt/_pillar/azureblob.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+:depends:
+    * `azure-storage-blob <https://pypi.org/project/azure-storage-blob/>`_ >= 12.0.0
+
+The Azure Blob external pillar can be configured with the following parameters:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - azureblob:
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:17]: give sanitized example config params

File: salt/_pillar/azureblob.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+from copy import deepcopy
+
+# Import 3rd-party libs
+# pylint: disable=import-error,no-name-in-module,redefined-builtin
+from salt.ext import six
+from salt.ext.six.moves import filter
+# pylint: enable=import-error,no-name-in-module,redefined-builtin
+
+# Import Azure libs
+HAS_LIBS = False
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:55]: `HAS_LIBS` should be used in a `__virtual__` function to be excluded from the loader if deps aren't present

File: salt/_pillar/azureblob.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+        # Iterate over the generator
+        while True:
+            try:
+                blob = next(blob_list)
+            except StopIteration:
+                break
+
+            log.debug('Raw blob attributes: %s', blob)
+
+            # Directories end with "/". Make them locally.
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: remove `Make them locally`. That was from my test script.

File: salt/_pillar/azureblob.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+            if blob.name.endswith('/'):
+                # Recurse into the directory
+                _walk_blobs(prefix=blob.name)
+
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Remove empty line

File: salt/_pillar/azureblob.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+            if container not in metadata[saltenv]:
+                metadata[saltenv][container] = []
+
+            metadata[saltenv][container].append(blob)
+
+    _walk_blobs(saltenv=environment)
+
+    # write the metadata to disk
+    if os.path.isfile(cache_file):
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:315]: Does s3 use the `os` library for this, or salt util/exec modules?

File: salt/_pillar/azureblob.py [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+
+    :param saltenv: Specifies which environment the container represents.
+
+    :param path: The path of the file in the container.
+
+    '''
+    env_meta = metadata[saltenv] if saltenv in metadata else {}
+    container_meta = env_meta[container] if container in env_meta else {}
+
+    log.error(container_meta)
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: `log.error` lines leftover from testing?

File: salt/_pillar/azureblob.py [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+    log.error('container_meta ^^')
+
+    files_meta = list(list(filter((lambda k: 'name' in k), container_meta)))
+
+    log.error(files_meta)
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: `log.error` lines leftover from testing?

File: salt/_pillar/azureblob.py [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+log = logging.getLogger(__name__)
+
+
+def ext_pillar(minion_id,
+               pillar,  # pylint: disable=W0613
+               container,
+               connection_string,
+               multiple_env=False,
+               environment='base',
+               prefix='',
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: We discussed `prefix` briefly. I think it was intended to namespace pillar data in an S3 bucket. A bucket ~= a blob, so the namespacing in a blob is by container. So, I think we can drop the prefix parameter here.

File: salt/_pillar/azureblob.py [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+               blob_cache_expire=30,
+               blob_sync_on_update=True):
+    '''
+    Execute a command and read the output as YAML
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Add param docs

File: salt/_pillar/azureblob.py [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+    '''
+    # normpath is needed to remove appended '/' if root is empty string.
+    pillar_dir = os.path.normpath(os.path.join(_get_cache_dir(), environment, container))
+
+    if prefix:
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Drop prefix

File: salt/_pillar/azureblob.py [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+        pillar_dir = os.path.normpath(os.path.join(pillar_dir, prefix))
+
+    if __opts__['pillar_roots'].get(environment, []) == [pillar_dir]:
+        return {}
+
+    metadata = _init(connection_string, container, multiple_env, environment, prefix, blob_cache_expire)
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Drop prefix

File: salt/_pillar/azureblob.py [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+    opts['ext_pillar'] = [x for x in opts['ext_pillar'] if 'azureblob' not in x]
+
+    pil = Pillar(opts, __grains__, minion_id, environment)
+
+    compiled_pillar = pil.compile_pillar(ext=False)
+
+    return compiled_pillar
+
+
+def _init(connection_string, container, multiple_env, environment, prefix, blob_cache_expire):
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Drop prefix

File: salt/_pillar/azureblob.py [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+    :param multiple_env: Specifies whether the pillar should interpret top level folders as pillar environments.
+        Defaults to false.
+
+    :param environment: Specifies which environment the container represents when in single environment mode. Defaults
+        to 'base' and is ignored if multiple_env is set as True.
+
+    :param blob_cache_expire: Specifies expiration time of the Azure Blob metadata cache file. Defaults to 30s.
+
+    '''
+    cache_file = _get_containers_cache_filename(container, prefix)
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Drop prefix

File: salt/_pillar/azureblob.py [hunk 15]
```diff
@@ -0,0 +1,449 @@
...
+    file_path = os.path.join(_get_cache_dir(), saltenv, container, path)
+
+    # make sure container and saltenv directories exist
+    if not os.path.exists(os.path.dirname(file_path)):
+        os.makedirs(os.path.dirname(file_path))
+
+    return file_path
+
+
+def _get_containers_cache_filename(container, prefix):
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Drop prefix

File: salt/_pillar/azureblob.py [hunk 16]
```diff
@@ -0,0 +1,449 @@
...
+    Return the filename of the cache for container contents. Create the path if it does not exist.
+
+    :param container: The name of the target Azure Blob Container.
+
+    '''
+    cache_dir = _get_cache_dir()
+    if not os.path.exists(cache_dir):
+        os.makedirs(cache_dir)
+
+    return os.path.join(cache_dir, '{0}-{1}-files.cache'.format(container, prefix))
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Drop prefix

File: salt/_pillar/azureblob.py [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+
+        # Create the ContainerClient object
+        container_client = blob_service_client.get_container_client(container)
+    except Exception as exc:
+        log.error('Exception: %s', exc)
+        return False
+
+    metadata = {}
+
+    def _walk_blobs(saltenv='base', prefix=None):
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:282]: Not this `prefix`... he's different

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]
Going to test functionality, but here are some things to change in the meantime.

File: salt/_pillar/azureblob.py (hunk 6 above)
Comment (Ajnbro) [reply salt/_pillar/azureblob.py:315]: The s3 ext_pillar use the os library for that. Those lines are actually copied directly from the s3 ext_pillar.

File: salt/_pillar/azureblob.py [hunk 18]
```diff
@@ -0,0 +1,460 @@
...
+# Import Salt libs
+from salt.pillar import Pillar
+import salt.utils.files
+import salt.utils.hashutils
+
+__virtualname__ = 'azureblob'
+
+# Set up logging
+log = logging.getLogger(__name__)
+
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:71]: 2 empty lines in between functions

File: salt/_pillar/azureblob.py [hunk 19]
```diff
@@ -0,0 +1,460 @@
...
+def __virtual__():
+    if not HAS_LIBS:
+        return (
+            False,
+            'The following dependency is required to use the Azure Blob ext_pillar: '
+            'Microsoft Storage Blob >= 12.0.0 '
+        )
+
+    return __virtualname__
+
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:82]: 2 empty lines in between functions

File: salt/_pillar/azureblob.py [hunk 20]
```diff
@@ -0,0 +1,460 @@
...
+            log.debug('Raw blob attributes: %s', blob)
+
+            # Directories end with "/".
+            if blob.name.endswith('/'):
+                # Recurse into the directory
+                _walk_blobs(prefix=blob.name)
+                continue
+
+            if multiple_env:
+                saltenv = ('base' if prefix == '.' else prefix)
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py]: Tested with `multiple_env`... we need to change this line to:
```
saltenv = ('base' if (not prefix or prefix == '.') else prefix[:-1])
```

File: salt/_pillar/azureblob.py [hunk 21]
```diff
@@ -0,0 +1,460 @@
...
+    :param blob_sync_on_update: Specifies if the cache is synced on update. Defaults to True.
+
+    '''
+    # normpath is needed to remove appended '/' if root is empty string.
+    pillar_dir = os.path.normpath(os.path.join(_get_cache_dir(), environment, container))
+
+    if __opts__['pillar_roots'].get(environment, []) == [pillar_dir]:
+        return {}
+
+    metadata = _init(connection_string, container, multiple_env, environment, blob_cache_expire)
```
Comment (nicholasmhughes) [inline salt/_pillar/azureblob.py:116]: During testing, it was helpful to have this line:
```
log.debug('Blob metadata: %s', metadata)
```

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]
After these changes, I think we're good to go.

**Review by nicholasmhughes — APPROVED** [review]
great work! :+1:

**Pull request merged**